include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp

```diff
@@ -51,291 +51,448 @@ namespace nil {
+                                    v += theta_acc * lookup_tag * plonk_columns.constant(l_table.lookup_options[o_id][i].index);
+                                    theta_acc *= theta;
+                                }
+                                v *= mask_assignment;
+                                auto reduced_v = reduce_dfs_polynomial_domain(v, basic_domain->m);
+                                lookup_value.push_back(v);
+                            }
+                        }
 
-                        math::polynomial_dfs<typename FieldType::value_type> F_compr_input(basic_domain->m - 1,
-                                                                                           basic_domain->m, 0);
+                        // Prepare lookup input                        
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> lookup_input;
+                        for( std::size_t g_id = 0; g_id < lookup_gates.size(); g_id++){
+                            auto &gate = lookup_gates[g_id];
+                            math::expression<DfsVariableType> expr; 
+                            math::polynomial_dfs<typename FieldType::value_type> lookup_selector = plonk_columns.selector(gate.tag_index);
+                            for( std::size_t c_id = 0; c_id < gate.constraints.size(); c_id++){
+                                auto &constraint = gate.constraints[c_id];
+                                math::polynomial_dfs<typename FieldType::value_type> l = lookup_selector * (typename FieldType::value_type(constraint.table_id));
+                                theta_acc = theta;
+                                for( std::size_t k = 0; k < constraint.lookup_input.size(); k++){
+                                    expr = converter.convert(constraint.lookup_input[k]);
+                                    math::cached_expression_evaluator<DfsVariableType> evaluator(expr, get_var_value);
+
+                                    l += theta_acc * lookup_selector * evaluator.evaluate();
+                                    theta_acc *= theta;
+                                }
+                               lookup_input.push_back(l);
+                            }
+                        }
+                        // Lookup_input and lookup_value are ready
+                        // Now sort them!
 
-                        math::polynomial_dfs<typename FieldType::value_type> F_compr_value(basic_domain->m - 1,
-                                                                                           basic_domain->m, 0);
+                        //  Reduce value and input
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> reduced_value;
+                        for( std::size_t i = 0; i < lookup_value.size(); i++ ){
+                            reduced_value.push_back(reduce_dfs_polynomial_domain(lookup_value[i], basic_domain->m));
```

Comment:
If we somehow divide this function into parts, we may save some memory. For example lookup_value variable is not used after this point, so it could be deleted already, if we exited some function here.

This function is pretty large, maybe this polynomial sorting part could be moved to a separate function?

Reply:
Removed sorting into separate function